### ISSUE TITLE
Cannot reference syslog-address by service name

### ISSUE THREAD

**Issue by lucj** [docker/docker#20370]
As suggested, I transfer the issue from https://github.com/docker/compose/issues/2657

When I test with a compose file like this one

    version: '2'
    services:
      #
      # Application services
      #
      app:
        image: hello-world
        depends_on:
          - logstash
        links:
          - logstash
        logging:
          driver: syslog
          options:
            syslog-address: "tcp://logstash:5514"
      #
      # ELK services
      #
      elasticsearch:
        image: 'XXX/elasticsearch:latest'
        ports:
          - '9200:9200'
        volumes:
          - elasticsearch-data:/usr/share/elasticsearch/data
      kibana:
        image: 'XXX/kibana:latest'
        links:
          - elasticsearch
        depends_on:
          - elasticsearch
        ports:
          - '5601:5601'
      logstash:
        image: 'XXX/logstash:latest'
        command: 'logstash -f /etc/logstash/conf.d/logstash.conf'
        links:
          - elasticsearch
        depends_on:
          - elasticsearch
        ports:
          - '5000:5000'
          - '5514:5514'
    volumes:
      elasticsearch-data:
        driver: local

I have the ‘ERROR: Failed to initialize logging driver: dial tcp: lookup logstash on 8.8.8.8:53: no such host’ error.

Seems that in the logging option of the app, the logstash service cannot be referenced in syslog-address: "tcp://logstash:5514".
Any idea of a workaround ?

**Comment by phemmer** [comment]
What version of docker are you using? This should be working on versions >= 10.0

**Comment by morallo** [comment]
Not working here with docker 1.10.1

**Comment by phemmer** [comment]
Hrm. Cause the error message `lookup logstash on 8.8.8.8:53: no such host` indicates that the container is using `8.8.8.8` as its nameserver. Docker 1.10 provides an embedded nameserver with address `127.0.0.11`. Are you overriding the name server or any dns related settings anywhere?

**Comment by thaJeztah** [comment]
@phemmer I suspect the logging options don't resolve the hostname from within the containers namespace, but from the host/daemon, but not sure

**Comment by morallo** [comment]
That's the explanation given in https://github.com/docker/compose/issues/2657 and https://github.com/docker/compose/issues/2182#issuecomment-147782211

**Comment by phemmer** [comment]
Oh, I misunderstood. I thought the error was being generated by the app inside the container, not by docker.

**Comment by lucj** [comment]
@phemmer sorry, I should have indicated this when filling the issue, I'm also using 1.10.1

**Comment by mavenugo** [comment]
Though this is an interesting use-case, It is not easy to achieve (unless we deploy special tricks).

Reason : Name resolution between containers (via container name, links, aliases) are handled by the embedded DNS server (or the expanded server) in the context of the logical network formed between the containers. These networks are in isolated space managed and handled by the network driver. In some cases (such as overlay driver), the isolated network space is not even accessible from the host network namespace directly. Since the docker daemon operates in the host network stack, it may not even be able to access the service exposed within an application network stack.
We could deploy some special tricks such as operating the daemon in container's network stack. But it is trickier and complex.

The simpler alternative is to do port-mapping and access the mapped service from the docker daemon on the host-networking stack.

**Comment by lucj** [comment]
Do you mean the workaround proposed by @dave-tucker ?
Will this work if logstash service and the app that needs to log to it are not deployed on the same host ?

**Comment by MBuffenoir** [comment]
I have been running in the same issue trying to run a similar compose file, this time on swarm (with a special network dedicated to logging). 
It felt like a good idea to leverage overlay network to spawn ELK in the same compose file than my app (not so sure on security point of view though ...)

**Comment by everett-toews** [comment]
Here's a simple example of the issue when trying to use overlay networks, ELK, and the GELF logging driver on a Swarm.

```yaml
version: '2'

services:
  elasticsearch:
    container_name: elasticsearch
    image: elasticsearch:2.3
    restart: unless-stopped
    networks:
      - logging
    expose:
      - 9200
      - 9300

  logstash:
    container_name: logstash
    image: logstash:2.3
    restart: unless-stopped
    networks:
      - logging
    expose:
      - "12201/udp"
    depends_on:
      - elasticsearch
    command: logstash -e 'input { gelf { type => docker port => 12201 } } output { elasticsearch { hosts => ["elasticsearch:9200"] } stdout { codec => rubydebug } }'

  kibana:
    container_name: kibana
    image: kibana:4.5
    restart: unless-stopped
    networks:
      - logging
    ports:
      - 5601
    depends_on:
      - elasticsearch
    environment:
      - ELASTICSEARCH_URL=http://elasticsearch:9200

  nginx:
    container_name: nginx
    image: nginx:1.9
    restart: unless-stopped
    networks:
      - logging
    ports:
      - "80:80"
    logging:
      driver: gelf
      options:
        gelf-address: "udp://logstash:12201"
    depends_on:
      - logstash

networks:
  logging: {}
```

Running it produces the error

```bash
$ docker-compose up
Creating elasticsearch
Creating logstash
Creating kibana
Creating nginx

ERROR: for nginx  500 Internal Server Error: Failed to initialize logging driver: gelf: cannot connect to GELF endpoint: logstash:12201 dial udp: lookup logstash on xxx.xxx.xxx.xxx:53: no such host
Attaching to elasticsearch, logstash, kibana
...
```

I've confirmed that the ELK portion is working using this [GELF Client](https://github.com/orionvm/python-gelfclient). It's the logging driver running on the host network stack as @mavenugo explained.

There are some solutions floating around that say to change `gelf-address: "udp://logstash:12201"` to `gelf-address: "udp://localhost:12201"` but that doesn't work with an overlay network. The error message does go away but no logs are actually written.

Relevant versions

```bash
$ docker version
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 21:49:11 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      swarm/1.2.0
 API version:  1.22
 Go version:   go1.5.4
 Git commit:   a6c1f14
 Built:        Wed Apr 13 05:58:31 UTC 2016
 OS/Arch:      linux/amd64

$ docker-compose version
docker-compose version 1.7.0, build 0d7bf73
docker-py version: 1.8.0
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1j 15 Oct 2014
```

**Comment by gprivitera** [comment]
Is there any update on the issue that @everett-toews brought up?

**Comment by thaJeztah** [comment]
@gprivitera you can still use log-drivers, however, sending logs to a _container_ won't work if the container is only accessible over a docker network (as in the example @everett-toews showed). Logs are collected by the _daemon_. The daemon itself is not part of the docker-network, so cannot resolve / directly connect to a container, unless the container is exposing/publishing a port. See https://github.com/docker/docker/issues/20370#issuecomment-185229591

**Comment by Hermain** [comment]
I am running a swarm with similar setup as @everett-toews with the difference that I publish port 12001. However I am still not able to forward logs with localhost from another machine in swarm and instead need to use 0.0.0.0. 
Any idea how to fix this?

**Comment by Hermain** [comment]
The solution was to publish ports:
      - "127.0.0.1:12201:12201/udp"
instead of only 12201:12201. My problem is solved. Now I only have the problem that conntrack cashes the connection even after the receiving container dies.

**Comment by robbatt** [comment]
I have the same problem. I am using fluentd, but this probably applies to syslogd and journald as well. Containers with fluentd logging-driver can not connect to a fluentd container within the same stack in swarm mode. 

As the fluentd architecture suggests you should have a fluentd proxy for the containers of your application, then "forward" it to a "global collector" that runs in front of elasticsearch and kibana. The recommended way to map the proxy to port 24224 of the host and let all containers log to this host port works, but it does not scale well.

If I want to run a stack on my dev machine it requires either a preinstalled fluentd daemon, or a free port 24224 which means I cannot run 2 stacks in parallel with default config. Both options are not desireable because they introduce an external dependency to my docker deployment. Workarounds with other ports require additional configuration and administration for every environment/swarm this stack is deployed to.

I'd love to see this working as a minimal configuration:

```
version: '3'
services:

  app:
    image: mycompany/app
    logging:
      driver: "fluentd"
      options:
        # this forwards logs to the own application stack, 
        # this should work once logger service is up (currently not supported)
        # would be even better if I could skip the default port
        fluentd-address: logger:24224

  logger:
    # listen to own port 24224
    # forwards to own stdout and host port 24224
    # if no daemon is running, the stack works anyway, the application is decoupled
    image: mycompany/fluentd-proxy
```

second stack in swarm might be the log analysis, which picks up logs from all stacks running in the swarm
```
version: '3'
services:

  logcollect:
    # listen to host port 24224
    # forwards to service elasticsearch
    image: mycompany/fluentd-collector

  elasticsearch:
    image: elasticsearch

  kibana:
    image: kibana
    ports:
      - "5601:5601"
```

**Comment by thaJeztah** [comment]
You cannot use the container-container network for logging; logging is handled by the docker daemon, not by the container; the container sends logs to `stdout` / `stderr`, and the docker daemon captures that output, and passes it to the logging driver. The daemon's networking (and the logging driver) are not part of the container-container network, thus cannot send the logs there.

**Comment by tangoromeo** [comment]
For some reason i'm not yet fully understaind this one is working:
    logging:
        driver: fluentd
        options:
          fluentd-address: :24224
          tag: translator

**Comment by thaJeztah** [comment]
@tangoromeo if the container that runs fluentd has its port _published_ on port 24224, then that port is publicly accessible on the host's external IP address, but also through `localhost:24224` / `127.0.0.1: 24224` (through hairpin NAT, or the userland-proxy). In that case, the daemon would be able to access the container through its exposed port.

Be aware though that that also means that _anyone_ could access that port (as it's publicly accessible). Which may not be what you want 😅

**Comment by thaJeztah** [comment]
@tangoromeo looks like the `fluentd` package uses defaults for any field that is not set; https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/vendor/github.com/fluent/fluent-logger-golang/fluent/fluent.go#L63-L100

**Comment by tangoromeo** [comment]
@thaJeztah that explains a lot :)
thanks!

**Comment by vikashsaini01** [comment]
> Though this is an interesting use-case, It is not easy to achieve (unless we deploy special tricks).
> 
> Reason : Name resolution between containers (via container name, links, aliases) are handled by the embedded DNS server (or the expanded server) in the context of the logical network formed between the containers. These networks are in isolated space managed and handled by the network driver. In some cases (such as overlay driver), the isolated network space is not even accessible from the host network namespace directly. Since the docker daemon operates in the host network stack, it may not even be able to access the service exposed within an application network stack.
> We could deploy some special tricks such as operating the daemon in container's network stack. But it is trickier and complex.
> 
> The simpler alternative is to do port-mapping and access the mapped service from the docker daemon on the host-networking stack.

On dockerCE for Windows it worked like
```
  demo:
    build: demo
    logging:
      driver: syslog
      options:
        syslog-address: tcp://host.docker.internal:5000  #-host.docker.internal resolved the port
    depends_on:
    - demo-logstash
    ports:
    - 8080:8080
```

**Comment by SimonPickup** [comment]
I have solved this for fluentd using UNIX domain sockets (https://stackoverflow.com/a/78357268/6426003). Presumably this can be adapted for syslog and others.
